File: easy/Two_Sum.c
/*
---------------------------------------------
Two_Sum
from : https://leetcode.com/problems/two-sum/
---------------------------------------------
*/

#include <string>
#include <vector>
#include <algorithm>

class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        vector<pair<int, int>> v;
        int i = 0, j;
        for(auto item : nums)
            v.push_back(make_pair(item, i++));
        sort(v.begin(), v.end());
        for(i = 0; i<nums.size() - 1; i++)
            for(j = i+1; j<nums.size() && v[i].first + v[j].first <= target; j++)
                if(v[i].first + v[j].first == target)
                    return {v[i].second, v[j].second};
        return {0, 0};
    }
};

File: PHP/Easy/Two_Sum.php
//from : https://leetcode.com/problems/two-sum/

<?php

class Solution {

    /**
     * @param Integer[] $nums
     * @param Integer $target
     * @return Integer[]
     */
    /* naive */
    /*
    function naive_twoSum($nums, $target) {
        foreach ($nums as $left_key=>$left_value) {
            foreach ($nums as $right_key=>$right_value) {
                if ($right_key <= $left_key) {
                        continue;
                } else if ($left_value + $right_value === $target) {
                        return array($left_key, $right_key);
                }
            }
        }
        
        
        
        //never reach this line
        return array(-1, -1);
    }
    */
    
    /* faster */
    function twoSum($nums, $target) {
        $array_of_candidates = [];
        foreach ($nums as $key => $value) {
            $candidate = ($target - $value);
            if (array_key_exists($candidate, $array_of_candidates)) {
                return array($array_of_candidates[$candidate], $key);
            } else {
                $array_of_candidates[$value] = $key;
            }
        }
        
        //never reach this line
        return array(-1, -1);
    }
}


?>
